Introduce zero overhead loop

martien-de-jong opened the pull request:
Final verdict: hwloop mostly causes significant PMsize expansion and frequent slight instruction count regressions (~ 10-100 cycles)

We have a few significant wins, e.g. GEMM_int8_1 InsnCount 55877 -> 53775
I didn't find any functional incorrectness and it is switched off by default. 

I would like to commit now, and propose a follow-up to take the loop size into account at e.g. legalization time, where it is relatively easy to allocate a virtual loopcount register. This on the basis that we don't gain on big loops, since there the loop body is dominated by memory, move and vector instructions.



This is a direct port of Abnikant's  original aie-private PR.
The representation of PseudoLoopEnd for analyzeBranch has changed significantly; We always push two components. The first is the opcode, the second the additional operand which can not be derived from the target block.

It should be noted that ZOL does not handle zero or negative loopcounts correctly. As such we need to establish that it is positive, e.g. by loop guarding or by interpreting pragma-like directives.

abnikant committed:
[AIE2] Legalize loop_decrement

martien-de-jong commented on llvm/lib/Target/AIE/AIELegalizerInfo.cpp:
```diff
@@ -1164,3 +1164,29 @@ bool AIELegalizerInfo::legalizeG_FADDSUB(LegalizerHelper &Helper,
+
+bool AIELegalizerInfo::legalizeIntrinsic(LegalizerHelper &Helper,
+                                         MachineInstr &MI) const {
+  switch (MI.getIntrinsicID()) {
```
There's a fix in a later commit that should go here.

martien-de-jong commented on llvm/lib/Target/AIE/AIEBaseInstrInfo.cpp:
```diff
@@ -78,7 +78,11 @@ void AIEBaseInstrInfo::insertNoop(MachineBasicBlock &MBB,
 bool AIEBaseInstrInfo::reverseBranchCondition(
     SmallVectorImpl<MachineOperand> &Cond) const {
-  assert((Cond.size() == 2) && "Invalid branch condition!");
+  assert((Cond.size() == 2 || Cond.size() == 3) && "Invalid branch condition!");
```
I think this is wrong. There's no opposite branchopcode for a hardware loop. Perhaps we can swap the branch targets, but it sounds iffy.

martien-de-jong replied:
This should not be called on a PseudoLoopEnd, which is represented by Cond.size() == 3. This change can be dropped, sanity assert will come from reversBranchCondition() which doesn't recognize PseudoLoopEnd

gbossu commented on llvm/lib/Target/AIE/AIE2AsmPrinter.cpp:
```diff
@@ -70,6 +70,39 @@ bool AIE2AsmPrinter::lowerOperand(const MachineOperand &MO,
+
+  // Insert a label before the instruction preceeding the PseudoLoopEnd.
+  if (MI->getOpcode() == AIE2::PseudoLoopEnd)
+    return;
```
Comment says we insert something, but this is actually an early exit?

gbossu commented on llvm/lib/Target/AIE/AIE2AsmPrinter.cpp:
```diff
@@ -70,6 +70,39 @@ bool AIE2AsmPrinter::lowerOperand(const MachineOperand &MO,
+    OutStreamer->getCommentOS() << "Loop end."
+                                << "\n";
+  }
+  AIEBaseAsmPrinter::emitInstruction(MI);
```
I'm a bit confused by this code. I think it tries to find a `PseudoLoopEnd`, and emit a label for the instruction before it.

Maybe change into something like:
```
 void AIE2AsmPrinter::emitInstruction(const MachineInstr *MI) {
  if (isPrecedingLoopEnd(MI)) {
    // Emit label and comment
  }
  AIEBaseAsmPrinter::emitInstruction(MI);
}
```

Also, I actually wonder if we should really trust the position of `PseudoLoopEnd`. I'm not sure it's guaranteed to really mark the end of the HW Loop with all the optimizations we run. Isn't the end of the loop always the end of the BB?

martien-de-jong replied:
Good question, PseudoLoopEnd is a terminator that lists both successors. I would hope it stays in place.

ehh, not true. It has a successor and the last-bundle symbol. But it is still a terminator.

martien-de-jong replied:
PseudoLoopEnd is a schedbarrier.

gbossu commented on llvm/lib/Target/AIE/AIE2InstrInfo.cpp:
```diff
@@ -940,7 +940,8 @@ bool AIE2InstrInfo::isLock(unsigned Opc) const {
 bool AIE2InstrInfo::isDelayedSchedBarrier(const MachineInstr &MI) const {
-  return MI.getOpcode() == AIE2::DelayedSchedBarrier;
+  return MI.getOpcode() == AIE2::DelayedSchedBarrier ||
+         MI.getOpcode() == AIE2::PseudoLoopEnd;
```
This would be better in the function below: `AIE2InstrInfo::isSchedBarrier`

martien-de-jong replied:
Fully agree. I think we're looking at a lot of ad-hoc adaptations.

gbossu commented on llvm/lib/Target/AIE/AIE2InstructionSelector.cpp:
```diff
@@ -214,6 +216,7 @@ class AIE2InstructionSelector : public InstructionSelector {
   // Select a control packet header stream write instruction from the
   // corresponding intrinsic.
   bool selectPutMS_CPH(MachineInstr &I, MachineRegisterInfo &MRI);
+  bool selectLoopDecrement(MachineInstr &I, MachineRegisterInfo &MRI);
```
I'd be happy to see .mir ISel tests for `LoopDecrement`, `LoopEnd`, `loop_iterations`

gbossu replied:
Should this commit maybe be squashed with [[AIE2] Clean up matching of ZOL End and add test](https://github.com/Xilinx/llvm-aie/pull/46/commits/a37cc925d1c2900944f875521ee80f0cb43939eb)?

gbossu commented on llvm/lib/Target/AIE/AIE2InstructionSelector.cpp:
```diff
@@ -959,9 +1001,8 @@ bool AIE2InstructionSelector::selectG_BRCOND(MachineInstr &I,
-  // resort to TableGen'ed selection patterns
-  return selectImpl(I, *CoverageInfo);
+  // select loop_decrement or resort to TableGen'ed selection patterns
+  return selectLoopDecrement(I, MRI);
```
I think something like `return selectLoopDecrement(I, MRI) || selectImpl(I, *CoverageInfo);` would be clearer

martien-de-jong replied:
And I don't like the fact that we feed a BRCOND as instruction into selectLoopDecrement, that's different from all other selectXXX functions.

martien-de-jong commented on llvm/lib/Target/AIE/AIE2InstructionSelector.cpp:
```diff
@@ -889,6 +897,40 @@ bool AIE2InstructionSelector::selectG_CONCAT_VECTORS(MachineInstr &I,
+  auto LoopDec = getDefIgnoringCopies(CondReg, MRI);
+  // Look for LoopDec Insn in the BB with G_BRCOND, the LoopDec is
+  // Selected else the loop break when it reaches first insn in the BB.
+  while (it != it_begin) {
```
@gbossu I don't understand this loop. Why not trust the getDefIgnoringCopies, do you see what is going on here?

martien-de-jong replied:
Loop could be removed without breaking existing tests

Commit:
[AIE] Several unrelated clang-tidies

Commit:
[AIE2] PseudoLoopEnd is a scheduler barrier

Also make PseudoLoopEnd a meta instruction to simplify emit logic

Make sure LoopStart/LoopEnd don't get duplicated in e.g. TailDuplication

Commit:
[AIE2] Mark loop registers as reserved

Commit:
[AIE] Late expansion of ZOL constructs

Completely remove empty ZOL

Commit:
[AIE] Harmonize analyzeBranch representation

PseudoLoopEnd is very similar to a regular conditional branch.
We need two Cond elements in order to reconstruct the instruction,
one is the opcode, the other is the condition register for JZ/JNZ and
the last-bundle label for PseudoLoopEnd

The operand order of PseudoLoopEnd was swapped to make it congruent to
the other conditonal branches

insertBranch needs to generate unconditional branch for FBB
even after PseudoLoopEnd.

Commit:
[AIE] Pad to 112 bytes after loop start

martien-de-jong commented on llvm/lib/Target/AIE/AIE2InstructionSelector.cpp:
```diff
@@ -897,38 +898,29 @@ bool AIE2InstructionSelector::selectG_CONCAT_VECTORS(MachineInstr &I,
+  if (LoopDecIntrinsic->getIntrinsicID() == Intrinsic::loop_decrement) {
+    MachineBasicBlock *DestMBB = BrCond.getOperand(1).getMBB();
+    MCContext &Context = BrCond.getParent()->getParent()->getContext();
+    MCSymbol *EndLabel = Context.createNamedTempSymbol("_112");
```
112: this label should be 112 bytes from the loopstart. We are not actually using it for that purpose yet.

gbossu commented on llvm/lib/Target/AIE/AIE2TargetTransformInfo.cpp:
```diff
@@ -57,7 +57,23 @@ bool AIE2TTIImpl::isHardwareLoopProfitable(Loop *L, ScalarEvolution &SE,
+    for (Instruction &I : *BB) {
+      if (isa<CallInst>(I) || isa<InvokeInst>(I)) {
+        if (const Function *F = cast<CallBase>(I).getCalledFunction()) {
+          if (!isLoweredToCall(F))
```
Curious: Where does `isLoweredToCall` come from? Is that a generic LLVM function?

martien-de-jong replied:
It's generic, with a note that it should be moved to a target-specific hook.

andcarminati replied:
I think we should have a custom implementation for this PR. See: 

```
void sum(double *a, double *b, double *c) {
    for(int i = 0; i < 30; i++) {
       c[i] = a[i] + b[i];
    }
}
```

It should walk in sync with GISel legalization rules for libcalls.

gbossu commented on llvm/test/CodeGen/AIE/aie2/hardware-loops/irtranslator-zol.ll:
```diff
@@ -0,0 +1,57 @@
+; NOTE: Assertions have been autogenerated by utils/update_mir_test_checks.py
+; RUN: llc -O2 -mtriple=aie2 -stop-after=irtranslator --enable-aie-hardware-loops --enable-aie-zero-overhead-loops %s -o - | FileCheck %s
+
```
Nit: Add a small description to tell about the test and what is expected.

gbossu commented on llvm/test/CodeGen/AIE/aie2/hardware-loops/nested.ll:
```diff
@@ -17,31 +17,19 @@ define void @nested(ptr nocapture %out, ptr nocapture readonly %in, i32 noundef
-; CHECK-NEXT:    nop
-; CHECK-NEXT:    nop
-; CHECK-NEXT:    mova r6, #0
-; CHECK-NEXT:    add.nc r5, r1, #-1
```
Do you understand the changes?

martien-de-jong replied:
As far as I can see, profitability of low overhead loops was reduced to single block loops. I removed the issue-limit=1, which probably wasn't very clever for this particular test.

martien-de-jong replied:
have restored issue-limit=1.

gbossu commented on llvm/lib/Target/AIE/AIELegalizerInfo.cpp:
```diff
@@ -1164,3 +1165,29 @@ bool AIELegalizerInfo::legalizeG_FADDSUB(LegalizerHelper &Helper,
+
+bool AIELegalizerInfo::legalizeIntrinsic(LegalizerHelper &Helper,
+                                         MachineInstr &MI) const {
+  switch (cast<GIntrinsic>(MI).getIntrinsicID()) {
```
Nit: Maybe leave a comment to say that other AIE-specific intrinsics are already created with legal operand types, and therefore do not need to be legalized here.

gbossu commented on llvm/test/CodeGen/AIE/aie2/hardware-loops/legalize-zol.mir:
```diff
@@ -38,9 +38,7 @@ body:             |
-  ; CHECK-NEXT:   [[C3:%[0-9]+]]:_(s32) = G_CONSTANT i32 1
-  ; CHECK-NEXT:   [[AND:%[0-9]+]]:_(s32) = G_AND [[ASSERT_ZEXT]], [[C3]]
-  ; CHECK-NEXT:   G_BRCOND [[AND]](s32), %bb.2
+  ; CHECK-NEXT:   G_BRCOND [[ASSERT_ZEXT]](s32), %bb.2
```
You did not change the legalizer, do you know why that test needed to be updated?
~Edit: I didn't see it's not your change. But I think it still makes sense to move that diff to the first commit if that's where it belongs.~

martien-de-jong replied:
I have assumed that the bit analysis has improved in upstream llvm, and now recognises that booleans are inrange.

gbossu commented on llvm/lib/Target/AIE/AIE2AsmPrinter.cpp:
```diff
@@ -70,6 +70,34 @@ bool AIE2AsmPrinter::lowerOperand(const MachineOperand &MO,
+    MCSymbol *S = LoopEnd->getOperand(1).getMCSymbol();
+    OutStreamer->emitLabel(S);
+    OutStreamer->getCommentOS() << "Loop end.\n";
+  }
```
Could you create a test for that change? Even if it's an end-to-end test, we could still use one as a baseline, and see that there is now a label.

gbossu commented on llvm/lib/Target/AIE/AIE2AsmPrinter.cpp:
```diff
@@ -70,6 +70,34 @@ bool AIE2AsmPrinter::lowerOperand(const MachineOperand &MO,
+void AIE2AsmPrinter::emitInstruction(const MachineInstr *MI) {
+  // PseudoLoopEnd is just here to carry some extra information. No need to
+  // emit anything
+  if (MI->getOpcode() == AIE2::PseudoLoopEnd)
```
Could we maybe just declare `PseudoLoopEnd` as a meta instruction to avoid the special case here?

martien-de-jong replied:
That worked, after reordering some condition logic.

gbossu commented on llvm/lib/Target/AIE/AIE2RegisterInfo.cpp:
```diff
@@ -108,6 +108,11 @@ BitVector AIE2RegisterInfo::getReservedRegs(const MachineFunction &MF) const {
+  // otherwise they are considered dead
+  markSuperRegs(Reserved, AIE2::LC);
+  markSuperRegs(Reserved, AIE2::LS);
+  markSuperRegs(Reserved, AIE2::LE);
```
That means we were deleting copies to those registers?

martien-de-jong replied:
It looks that way. Didn't check, but stands to reason.

gbossu commented on llvm/lib/Target/AIE/AIEBaseHardwareLoops.cpp:
```diff
@@ -164,8 +169,15 @@ bool AIEBaseHardwareLoops::ProcessLoop(MachineLoop *ML) {
 
-  LowOverheadLoop LoLoop(*ML, *MLI, *TRI, *TII);
+  MachineBasicBlock *LastMBB = ML->findLoopControlBlock();
+  // More than one exit, don't generate ZOL.
```
Can that happen?

martien-de-jong replied:
That is probably a remote possibility involving introducing a more-than-two way branch.

andcarminati commented:
Hi, the following code can cause some problem in this PR:

```
target datalayout = "e-m:e-p:20:32-i1:8:32-i8:8:32-i16:16:32-i32:32:32-f32:32:32-i64:32-f64:32-a:0:32-n32"
target triple = "aie2"

define i32 @main() {
entry:
  br label %for.body

for.body:                                  ; preds = %for.body, %entry
  %i = phi i16 [ 0, %entry ], [ 1, %for.body ]
  %cmp = icmp ult i16 %i, 1
  br i1 %cmp, label %for.body, label %if.end

if.end:                                     ; preds = %for.body
  ret i32 0
}
```

With:

```
llc --march=aie2 sample.ll  --enable-aie-hardware-loops  --enable-aie-zero-overhead-loops
```

Gives: 

```
llc: ../llvm/include/llvm/ADT/ilist_iterator.h:138: llvm::ilist_iterator::reference llvm::ilist_iterator<llvm::ilist_detail::node_options<llvm::MachineInstr, true, true, void, false>, false, false>::operator*() const [OptionsT = llvm::ilist_detail::node_options<llvm::MachineInstr, true, true, void, false>, IsReverse = false, IsConst = false]: Assertion `!NodePtr->isKnownSentinel()' failed.
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace.
Stack dump:
0.      Program arguments: ./bin/llc --march=aie2 reduced.ll --enable-aie-hardware-loops --enable-aie-zero-overhead-loops
1.      Running pass 'Function Pass Manager' on module 'reduced.ll'.
2.      Running pass 'InstructionSelect' on function '@main'
 #0 0x0000000005d38f97 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) ./../llvm/lib/Support/Unix/Signals.inc:723:13
 #1 0x0000000005d36f70 llvm::sys::RunSignalHandlers() ./../llvm/lib/Support/Signals.cpp:106:18
 #2 0x0000000005d3964a SignalHandler(int) ./../llvm/lib/Support/Unix/Signals.inc:413:1
 #3 0x00007f6f6c1a5520 (/lib/x86_64-linux-gnu/libc.so.6+0x42520)
 #4 0x00007f6f6c1f99fc __pthread_kill_implementation ./nptl/pthread_kill.c:44:76
 #5 0x00007f6f6c1f99fc __pthread_kill_internal ./nptl/pthread_kill.c:78:10
 #6 0x00007f6f6c1f99fc pthread_kill ./nptl/pthread_kill.c:89:10
 #7 0x00007f6f6c1a5476 gsignal ./signal/../sysdeps/posix/raise.c:27:6
 #8 0x00007f6f6c18b7f3 abort ./stdlib/abort.c:81:7
 #9 0x00007f6f6c18b71b _nl_load_domain ./intl/loadmsgcat.c:1177:9
#10 0x00007f6f6c19ce96 (/lib/x86_64-linux-gnu/libc.so.6+0x39e96)
```

andcarminati commented on llvm/lib/Target/AIE/AIE2InstructionSelector.cpp:
```diff
@@ -889,6 +898,31 @@ bool AIE2InstructionSelector::selectG_CONCAT_VECTORS(MachineInstr &I,
+    MCContext &Context = BrCond.getParent()->getParent()->getContext();
+    MCSymbol *EndLabel = Context.createNamedTempSymbol("_112");
+    MIB.buildInstr(AIE2::PseudoLoopEnd).addMBB(DestMBB).addSym(EndLabel);
+    LoopDec->eraseFromParent();
```
After the second erase, iterator got corrupted.

andcarminati commented:
Hi @martien-de-jong , another interesting case (sample.ll):

```
target datalayout = "e-m:e-p:20:32-i1:8:32-i8:8:32-i16:16:32-i32:32:32-f32:32:32-i64:32-f64:32-a:0:32-n32"
target triple = "aie2"

define i32 @main() {
entry:
  br label %for.body

for.body:                                     ; preds = %for.body, %entry
  %i = phi i32 [ 0, %entry ], [ %inc, %for.body ]
  %inc = add i32 %i, 1
  %exitcond = icmp eq i32 %inc, 0
  br i1 %exitcond, label %label, label %for.body

label:                                     ; preds = %label, %for.body
  br label %label
}
```

However, you need the following options (Elf emission, no loop scheduling):

```
llc --march=aie2 sample.ll  --enable-aie-hardware-loops  --enable-aie-zero-overhead-loops -aie-loop-aware=false -filetype=obj
```

Result:
```
<unknown>:0: error: Undefined temporary symbol .L_1120
```

Commit:
[AIE2] Instruction-select ZOL end

Commit:
[AIE] emit last-bundle symbol for ZOL

lower symbol in MC lowering

andcarminati commented on llvm/lib/Target/AIE/AIEFinalizeBundle.cpp:
```diff
@@ -20,8 +20,17 @@ using namespace llvm;
-  if (MI->isMetaInstruction() || MII->isBundled())
+  // LoopStart should be lowered before reaching here
+  assert(!TII->isHardwareLoopStart(MI->getOpcode()) &&
+         "LoopStart is not lowered \n\r");
```
Hi @martien-de-jong,

I can trigger this assert with the following code:

```
target datalayout = "e-m:e-p:20:32-i1:8:32-i8:8:32-i16:16:32-i32:32:32-f32:32:32-i64:32-f64:32-a:0:32-n32"
target triple = "aie2"

define i32 @main(i1 %cmp) {
entry:
  %call = tail call i32 null(ptr null, ptr null)
  br i1 %cmp, label %if.else, label %for.cond.preheader

for.cond.preheader:                          ; preds = %entry
  store i8 0, ptr null, align 4
  br label %for.body

if.else:                                      ; preds = %entry
  %bf = load i16, ptr null, align 4
  br label %for.body

exit:                        ; preds = %for.body
  ret i32 0

for.body:                                  ; preds = %for.body, %if.else, %for.cond.preheader
  %j = phi i32 [ %inc, %for.body ], [ 0, %if.else ], [ 0, %for.cond.preheader ]
  %0 = trunc i32 %j to i20
  %inc = add i32 %j, 1
  %exitcond = icmp eq i32 %inc, 0
  br i1 %exitcond, label %exit, label %for.body
}

```

martien-de-jong replied:
I'm on it. Probably an assumption on seeing LoopStart before LoopEnd.

martien-de-jong replied:
The true story: tail duplication copied LoopStart out into predecessor blocks. This scares hwloop expansion.

andcarminati commented on llvm/lib/Target/AIE/AIEBaseHardwareLoops.cpp:
```diff
@@ -206,48 +279,73 @@ bool AIEBaseHardwareLoops::ProcessLoop(MachineLoop *ML) {
 
-  // Future: Handle true hardware loop setup code
-  llvm_unreachable("Implement true hardware loop support");
+  if (LoLoop.isZeroOverheadLoop()) {
```
Do we really need to test this condition? Is there any situation where this if will evaluate to false?

martien-de-jong replied:
I'll check. Modify it to an assert if it can be.

andcarminati commented on llvm/lib/Target/AIE/AIEBaseInstrInfo.cpp:
```diff
@@ -267,17 +276,20 @@ unsigned AIEBaseInstrInfo::insertBranch(
+  // PseudoLoopEnd, the condition register for JZ/JNZ
   unsigned Opc = Cond[0].getImm();
-  MachineInstr &CondMI = *BuildMI(&MBB, DL, get(Opc)).add(Cond[1]).addMBB(TBB);
+  MachineInstrBuilder CBB = BuildMI(&MBB, DL, get(Opc));
```
As we already have TBB, FBB as BB, CBB seems to be a BB as well because of the name. Maybe a different name is enough to state that this is an instr.

gbossu commented on llvm/lib/Target/AIE/AIE2InstrInfo.td:
```diff
@@ -557,7 +557,7 @@ def LoopJNZ :
 
 // Zero overhead loops
-let hasSideEffects = 1, mayLoad = 0, mayStore = 0 in
+let hasSideEffects = true, mayLoad = 0, mayStore = 0, isNotDuplicable = true in
```
Super nit: While you're at it, also change the `0`s to `false`?

martien-de-jong replied:
Yes!

gbossu commented on llvm/test/CodeGen/AIE/aie2/hardware-loops/legalize-zol.mir:
```diff
@@ -0,0 +1,76 @@
+    %13:_(s32) = G_CONSTANT i32 1
+    %16:_(s32) = G_CONSTANT i32 0
+    %3:_(s32) = G_LOAD %0(p0) :: (load (s32))
+    G_INTRINSIC_W_SIDE_EFFECTS intrinsic(@llvm.set.loop.iterations), %2(s32)
```
Btw what are the semantics of `llvm.set.loop.iterations`? Is it always assumed to sit in a BB with a single successor? And in this case this gives the number of iterations for that successor?

martien-de-jong replied:
It is not duplicable, and it is placed in the preheader of the loop. I don't find much specified about nesting, or 1-1 use of setup and decrement. I assume that a set.loop.iterations will effectively push a loop level, decrement will maintain that loop level until the iterations run out, then it will pop. I think AIE could handle multiple set.loop.iterations for the same loop, but needs a single endloop.

gbossu commented on llvm/lib/Target/AIE/AIE2InstructionSelector.cpp:
```diff
@@ -214,6 +215,8 @@ class AIE2InstructionSelector : public InstructionSelector {
   // corresponding intrinsic.
   bool selectPutMS_CPH(MachineInstr &I, MachineRegisterInfo &MRI);
+  bool selectBrCondLoopDecrement(MachineInstr &BrCond,
+                                 MachineRegisterInfo &MRI);
```
Nit: Commit says it's cleaning something up, but I think it's only adding things now.

martien-de-jong replied:
Will fix.

gbossu commented on llvm/lib/Target/AIE/AIE2InstructionSelector.cpp:
```diff
@@ -889,6 +897,31 @@ bool AIE2InstructionSelector::selectG_CONCAT_VECTORS(MachineInstr &I,
+// Try to match BRCOND(Intrinsic::loop_decrement)
+bool AIE2InstructionSelector::selectBrCondLoopDecrement(
+    MachineInstr &BrCond, MachineRegisterInfo &MRI) {
+  MachineOperand &MO = BrCond.getOperand(0);
```
Super-nit: Maybe have an assert to verify the `BrCond.getOpcode() == G_BRCOND` precoindition?

gbossu commented on llvm/lib/Target/AIE/AIE2InstructionSelector.cpp:
```diff
@@ -889,6 +897,31 @@ bool AIE2InstructionSelector::selectG_CONCAT_VECTORS(MachineInstr &I,
+  Register CondReg = MO.getReg();
+
+  // Check if the condition is a LoopDecrement
+  auto *LoopDec = getDefIgnoringCopies(CondReg, MRI);
```
Can it actually happen that `LoopDec == nullptr`?

martien-de-jong replied:
I think it's very unlikely, CondReg would need to have an invalid type.

gbossu commented on llvm/test/CodeGen/AIE/aie2/hardware-loops/instruction-select-zol.mir:
```diff
@@ -0,0 +1,73 @@
+    %19:gprregbank(s32) = G_INTRINSIC_W_SIDE_EFFECTS intrinsic(@llvm.loop.decrement), %3(s32)
+    %20:gprregbank(s32) = G_ASSERT_ZEXT %19, 1
+    G_BRCOND %20(s32), %bb.2
+    G_BR %bb.1
```
I think that test could be simplified/split to really only focus on:
 - Isel for `llvm.set.loop.iterations`
 - ISel for `llvm.loop.decrement`

martien-de-jong replied:
ok.

gbossu commented on llvm/lib/Target/AIE/AIE2AsmPrinter.cpp:
```diff
@@ -70,6 +70,34 @@ bool AIE2AsmPrinter::lowerOperand(const MachineOperand &MO,
+    return;
+
+  // Check whether there's a PseudoLoopEnd following. If so, emit the
+  // last-bundle-label that it carries to designate this MI as the last bundle.
```
Note: The AsmPrinter has a notion of "post-instruction symbol", and it can trigger the emission of a symbol after an instruction is emitted. See:
```
      // If there is a post-instruction symbol, emit a label for it here.
      if (MCSymbol *S = MI.getPostInstrSymbol())
        OutStreamer->emitLabel(S);
```

Maybe we could hook into this mechanism?

gbossu replied:
And even:
```
      // If there is a pre-instruction symbol, emit a label for it here.
      if (MCSymbol *S = MI.getPreInstrSymbol())
        OutStreamer->emitLabel(S);
```
before emitting the instruction

martien-de-jong replied:
Sounds attractive.

gbossu commented on llvm/lib/Target/AIE/AIE2AsmPrinter.cpp:
```diff
@@ -70,6 +70,34 @@ bool AIE2AsmPrinter::lowerOperand(const MachineOperand &MO,
+      break;
+    }
+    Prev = &MII;
+  }
```
I think that analysis could be done a single time before emitting any instruction is called. If we can't use pre/post emission symbols, maybe move that analysis into e.g. `AIEBaseAsmPrinter::emitFunctionBodyStart`?

martien-de-jong replied:
Yep. Works like a charm.

gbossu commented on llvm/test/CodeGen/AIE/aie2/hardware-loops/loop-with-call.ll:
```diff
@@ -0,0 +1,132 @@
+; (c) Copyright 2024 Advanced Micro Devices, Inc. or its affiliates
+
+; RUN: llc -O2 -mtriple=aie2 --enable-aie-hardware-loops \
+; RUN:   --enable-aie-zero-overhead-loops %s -o - | FileCheck %s
```
I'd be happy to have those tests added earlier as baseline, so we can see how they get updated with each commit.

gbossu commented on llvm/lib/Target/AIE/AIEFinalizeBundle.cpp:
```diff
@@ -20,8 +20,17 @@ using namespace llvm;
+  assert(!TII->isHardwareLoopStart(MI->getOpcode()) &&
+         "LoopStart is not lowered \n\r");
+  if (MI->isMetaInstruction() || MII->isBundled() ||
+      TII->isHardwareLoopEnd(MI->getOpcode()))
```
Can you add a test for the alignment after LoopStart?

martien-de-jong replied:
Perhaps, but we don't need alignment. we piggy-back padding onto the alignment mechanism.

martien-de-jong replied:
I think there is already a test for the padding added in the bundle alignment PR. But I will check.

martien-de-jong replied:
Indeed, in CodeGen/AIE/aie2/elongate/*

gbossu replied:
But is there one specifically for HW Loops and the handling of "loop setup instructions"?

martien-de-jong replied:
yes

gbossu commented on llvm/lib/Target/AIE/AIEBaseAsmPrinter.cpp:
```diff
@@ -61,14 +61,32 @@ void AIEBaseAsmPrinter::emitFunctionBodyStart() {
+
+    // Scan for endloops to propagate the last-bundle label.
+    // The loop is separated from the previous one because it needs a bundle
+    // iterator instead of an instruction iterator.
```
That turned out pretty nice!

martien-de-jong replied:
Yes, sweet and lovely.

gbossu commented on llvm/test/CodeGen/AIE/aie2/hardware-loops/endlabel.mir:
```diff
@@ -0,0 +1,34 @@
+# NOTE: Assertions have been autogenerated by utils/update_mir_test_checks.py
```
Liar 😏

gbossu commented on llvm/test/CodeGen/AIE/aie2/hardware-loops/noduplication.mir:
```diff
@@ -0,0 +1,152 @@
+# NOTE: Assertions have been autogenerated by utils/update_mir_test_checks.py UTC_ARGS: --version 4
+# RUN: llc -mtriple=aie2 --start-after=instruction-select \
+# RUN:   --stop-before=aie-finalize-mi-bundles %s -o - | FileCheck %s
```
Super-Nit: I'm changing the scheduler so it outputs "correct" Bundles. Could you change that line to `--stop-after=aie-finalize-mi-bundles`? This way there won't be test updates.

gbossu commented on llvm/test/CodeGen/AIE/aie2/hardware-loops/complex-flow.mir:
```diff
@@ -0,0 +1,205 @@
+
+# This stresses block placement to a point where a ZOL doesn't fall-through
+# to its exit point. This gives rise to insertBranch of a PseudoLoopEnd with
+# a non-null FBB, which should generate an unconditional branch
```
Nit: I'd be happy with an ascii-art CFG :)

gbossu commented on llvm/lib/Target/AIE/AIEBaseInstrInfo.cpp:
```diff
@@ -172,8 +180,11 @@ bool AIEBaseInstrInfo::analyzeBranch(MachineBasicBlock &MBB,
+      // The basic block target is provided in the interfaces so we don't need
+      // to push that
       Cond.push_back(MachineOperand::CreateImm(I->getOpcode()));
-      Cond.push_back(I->getOperand(0));
```
Did that hurt to keep `Cond.push_back(I->getOperand(0));`? I'm still struggling to understand what kind of API `analyzeBranch` has.

martien-de-jong replied:
The api is that the target can push whatever it needs to reconstruct/invert a branch. We don't actually need that third operand, and I like occam's razor.

gbossu commented on llvm/lib/Target/AIE/AIEBaseInstrInfo.cpp:
```diff
@@ -269,17 +278,20 @@ unsigned AIEBaseInstrInfo::insertBranch(
+  if (isHardwareLoopEnd(Opc)) {
+    CBranchBuilder.addMBB(TBB).add(Cond[1]);
+  } else {
+    CBranchBuilder.add(Cond[1]).addMBB(TBB);
```
Super-nit: Maybe we could define `PseudoLoopEnd` to have the same operand order as other branches?

martien-de-jong replied:
Deal!

gbossu commented on llvm/lib/Target/AIE/AIEBaseInstrInfo.cpp:
```diff
@@ -184,18 +195,18 @@ bool AIEBaseInstrInfo::analyzeBranch(MachineBasicBlock &MBB,
-      else
-        parseCondBranch(*std::prev(I), TBB, Cond);
+      parseCondBranch(*Prev, TBB, Cond);
+    }
```
Nit: maybe let `parseCondBranch` determine whether it can actually parse the branch? E.g.
```
else if (!parseCondBranch(*Prev, TBB, Cond)) {
  return Unhandled;
}
```

gbossu commented on llvm/test/CodeGen/AIE/aie2/hardware-loops/zol-loop.mir:
```diff
@@ -0,0 +1,82 @@
+#
+# (c) Copyright 2024 Advanced Micro Devices, Inc. or its affiliates
+
+# RUN: llc -mtriple aie2 -run-pass=aie-hardware-loops \
```
Can you move that test to the previous commit where you add the code for `LoopStart` expansion?

gbossu reviewed (commented):
I'm done with the review. I think it looks good! Please go through the remaining comments, I'd be happy if some tests are moved around, but it's not such a big deal :)

andcarminati reviewed (dismissed):
LGTM. Nice work, credits to you and the origin author.

Commit:
[AIE] Update HW-Loop Profitability

konstantinschwarz commented on llvm/test/CodeGen/AIE/aie2/hardware-loops/instruction-select-zol-end.mir:
```diff
@@ -0,0 +1,27 @@
+# NOTE: Assertions have been autogenerated by utils/update_mir_test_checks.py
+# RUN: llc -O2 -mtriple=aie2 -run-pass=instruction-select %s -verify-machineinstrs -o - | FileCheck %s
+
+---
```
A couple of tests here are missing the license & copyright header. Could you please add those in a fixup PR @martien-de-jong?
